### PR TITLE
Cantera flamelet and tabulation capabilities added

### PR DESCRIPTION
**Pull request by axbriones**
Please fill in the issue number this pull request is fixing:
The development was performed on Cantera 2.3 but I think it can be integrated in the Master.

Fixes #

Changes proposed in this pull request:
- The Cantera code was modified to handle S-curve calculations of opposed flows.
- Unity Lewis number capability was also added.
- Energy equation source terms can be activated/deactivated from the python script.
- Many python scripts are added to perform FPI and FPV turbulence-chemistry interactions.
- This tabulations can be used to run CFD simulations of reacting flows.

### CODE REVIEW

**Comment by wandadars** [comment]
This is fantastic that you have done all this work. Flamelet generation in Cantera is a topic that I've personally been trying to work towards. This is lightyears ahead of the progress that I have made.

The huge commits are going to be an issue.

File: include/cantera/oneD/Domain1D.h
Comment (bryanwweber) [inline include/cantera/oneD/Domain1D.h:555]: Something went wrong with your indentation here. Please use spaces, not tabs, to indent your code.

File: include/cantera/oneD/StFlow.h
Comment (bryanwweber) [inline include/cantera/oneD/StFlow.h:23]: Please do not add your name to comments in the code.

File: include/cantera/oneD/StFlow.h
Comment (bryanwweber) [inline include/cantera/oneD/StFlow.h:90]: It looks like this code is not based on the most current revision of the master branch. Please rebase onto at least 66ba4b5b507fd182a35f0b3310d22fe9ee72e51c and fix any merge conflicts. You're actually removing things that have been specifically added here.

File: include/cantera/oneD/StFlow.h
Comment (bryanwweber) [inline include/cantera/oneD/StFlow.h:245]: Same as before about spacing

File: include/cantera/oneD/StFlow.h
Comment (bryanwweber) [inline include/cantera/oneD/StFlow.h:325]: Please add useful comments. Extra variables in extra equations won't help anyone else in the future.

File: interfaces/cython/cantera/examples/onedim/ConvoluteCpp.h
Comment (bryanwweber) [inline interfaces/cython/cantera/examples/onedim/ConvoluteCpp.h:36]: Documentation should be provided in the format of Doxygen docstrings. Please see the other C++/.h files for examples

File: interfaces/cython/cantera/onedim.pyx
Comment (bryanwweber) [inline interfaces/cython/cantera/onedim.pyx:1134]: In function definitions, there should not be spaces between keyword arguments and the `=` sign

File: src/oneD/StFlow.cpp
Comment (bryanwweber) [inline src/oneD/StFlow.cpp:106]: This comment is not needed, the commit logs will show the history of the changes, and with useful commit messages, the reasons will be clear as well.

**Comment by wandadars** [comment]
Where did you add the unity lewis number transport capability?

**Comment by wandadars** [comment]
What does "handle S-curve calculations" mean? Numerical continuation?

File: interfaces/cython/cantera/examples/onedim/convolute.pyx
Comment (bryanwweber) [inline interfaces/cython/cantera/examples/onedim/convolute.pyx:103]: There should be 2 spaces between the code and the inline comment

    cdef CreateOutputFile c_conv  # hold a C++ instance which we're wrapping
                                ^^

but is this comment needed?

File: interfaces/cython/cantera/examples/onedim/UnscaleVariances.py
Comment (bryanwweber) [inline interfaces/cython/cantera/examples/onedim/UnscaleVariances.py:108]: Python functions should be documented with docstrings, not comments. Also, there are too many blank lines here (one is sufficient) - please avoid adding extraneous whitespace

File: interfaces/cython/cantera/examples/onedim/UnscaleVariances.py
Comment (bryanwweber) [inline interfaces/cython/cantera/examples/onedim/UnscaleVariances.py:59]: Please do not leave open file handles like this. The scope of the input file handle should be limited such that it can fit in a context block:

    with open (fileName, 'r'):
        # read the file

If needed, you can store the file name itself as an instance variable, or the data read from the file. In addition, variables should be named with underscores, rather than camelCase:

    self._fileName -> self._file_name or self._filename

File: interfaces/cython/cantera/examples/onedim/UnscaleVariances.py
Comment (bryanwweber) [inline interfaces/cython/cantera/examples/onedim/UnscaleVariances.py:184]: Same comment here about file handles. All of the output can be collected in, e.g., a list and then written to a file in a single context block.

**Comment by wandadars** [comment]
I see that many of the Python scripts for the flamelet table generation are placed in the example directory. I think it would make sense to have a subdirectory just for the flamelet stuff due to how many different Python files are associated with that capability. Currently each Python script in the examples directory correlates to an isolated example(that's my current understanding).

File: interfaces/cython/cantera/examples/onedim/ConvoluteCpp.h
Comment (bryanwweber) [inline interfaces/cython/cantera/examples/onedim/ConvoluteCpp.h:2]: Please note that the copyright cannot be assigned to your university. Cantera is licensed under the BSD 3-clause license and all contributions must be licensed under the same thing.

**Review by bryanwweber — CHANGES_REQUESTED** [review]
Hi! Thank you for this contribution! Many of the ideas here I think will be useful to include, but I've made several comments about the code style below that will definitely need to be fixed before we can merge this.

There are several places where code has been changed between 2.3.0 and the current master, such that you will need to rebase this code onto the master branch to incorporate those changes. Moreover, it would be much better if you can split each commit into several commits and explain what each commit changes in the commit message. This will make it easier to track the changes in the future.

For the examples, I'm not sure what the best place for those is. I think we try to avoid having examples with many internal dependencies between modules (for instance, your `Utils` module), and focus on demonstrating the basic capabilities of Cantera. Your use here seems like a more applied example, which may be more appropriate in our Jupyter Notebook repository, or perhaps another place. This will have to take some discussion to figure out. Nonetheless, I made a few style comments on the Python code as well.

Finally, the biggest roadblock is the licensing. As I noted in a comment, code included in Cantera must be licensed under the BSD 3-clause license. The copyright header you have makes it so that people cannot download even the source code without checking with the University of Dayton. This is clearly not possible for a project such as Cantera, so this will have to change.

Again, thank you for taking the time to make this contribution! I'm sure other folks will be around soon-ish to give more thorough feedback (and I might take some more time later to do that as well), but I wanted to give you something quickly since I had some time this afternoon.

**Comment by axbriones** [comment]
@wandadars 
The Unity number capability is passed from the python script (diffusion_flame_s_curve_retry.py or ) and it is passed all the way to the source code. I believe to StFlow.cpp and/or Sim1D.cpp.

**Comment by bryanwweber** [comment]
@axbriones A few other thoughts:

1. Check out the contributing guide for more about the style: https://github.com/Cantera/cantera/blob/master/CONTRIBUTING.md
2. If you need help, particularly with any of the git/rebasing stuff, I would be happy to try to help.

**Comment by axbriones** [comment]
@wandadars 
S-curve refers to numerical continuation

**Comment by axbriones** [comment]
@wandadars 
Yes, I agree that these python scripts should be a separate folder. I am new at GitHub I don't really know how to use it yet.

File: include/cantera/oneD/Domain1D.h
Comment (axbriones) [reply include/cantera/oneD/Domain1D.h:555]: Thanks for the comment I will look into it.

File: include/cantera/oneD/StFlow.h
Comment (axbriones) [reply include/cantera/oneD/StFlow.h:23]: Sorry, we left these name when didn't realize. Would you be making any changes or should I need to address all the comments?

**Comment by wandadars** [comment]
@axbriones Welcome to the Cantera Github page!  We have a Gitter channel where anyone is free to chat. It might be helpful to pop on there if you need some help using github. This thread is also fine too.

File: include/cantera/oneD/Domain1D.h
Comment (axbriones) [reply include/cantera/oneD/Domain1D.h:555]: I will get back to you about the University of Dayton Copyright statement next week.

File: include/cantera/oneD/StFlow.h
Comment (bryanwweber) [reply include/cantera/oneD/StFlow.h:23]: @axbriones You should address the comments. We can give advice about the best way to update all of this.

Comment (wandadars) [reply include/cantera/oneD/StFlow.h]: @axbriones The method to address the changes is:
1.) Clone your repo to your local computer: git clone https://github.com/axbriones/cantera.git
2.) Make the changes that are requested on your master branch(this will be the default branch when you clone). 
    Try to make 1 change at a time. 
    Once you've finished an adjustment for a single change request, do a:
                      git add -u   and 
                     git commit -m "I have made these changes"
3.) Follow step number 2 for all changes
4.) Do a "git push origin master"
5.) These changes will immediately be reflected in this pull request because github auto-updates pull requests with any changes made to the branch under consideration.
6.) Repeat this process as additional change requests are made.

@bryanwweber Would you add anything to this? @axbriones Is a new Github user.

**Comment by wandadars** [comment]
@axbriones The method to address the requested changes is:
1.) Clone your repo to your local computer: git clone https://github.com/axbriones/cantera.git
2.) Make the changes that are requested on your master branch(this will be the default branch when you clone).
Try to make 1 change at a time.
Once you've finished an adjustment for a single change request, do a:
git add <changed_file> and
git commit -m "I have made these changes"
3.) Follow step number 2 for all changes
4.) Do a "git push origin master"
5.) These changes will immediately be reflected in this pull request because github auto-updates pull requests with any changes made to the branch under consideration.
6.) Repeat this process as additional change requests are made.

@bryanwweber Would you add anything to this? @axbriones Is a new Github user.

File: interfaces/cython/cantera/examples/onedim/ConvoluteCpp.h
Comment (bryanwweber) [reply interfaces/cython/cantera/examples/onedim/ConvoluteCpp.h:2]: But please note that you should add yourself to the `AUTHORS` file

**Comment by bryanwweber** [comment]
@wandadars No, I don't think that will be sufficient for this case, since the code is based on Cantera 2.3.0, it reverts all of the changes since then in the affected files. I suspect it will be easier to start from scratch, cloning the Cantera/cantera master branch and re-applying these changes to a separate branch from master and probably creating a new pull request. Your instructions for commiting are mostly correct, although I'm not sure why you're adding the `-u` flag, which prevents adding new files that aren't already in the repository.

Regardless, we need to sort out the licensing first, so not much reason to really sort the commit history out until then.

**Comment by bryanwweber** [comment]
@axbriones You mentioned to me offline that the copyright was sorted out. Can you clarify what that means from your end? Thanks!

**Comment by axbriones** [comment]
Bryan,



From the University of Dayton standpoint its is okay if I remove the copyright statement. However, I still have to take care of another legal aspect before sharing the entire code freely with the community. This work is in progress and requires some approval that I am waiting for. Once, this is complete I will proceed to the commits and pull request to merge or version and our python scripts with your Cantera 2.4



Thanks.



Alex

________________________________

**Comment by bryanwweber** [comment]
@axbriones That is great to hear. We can certainly put a citation for the work somewhere in the documentation, and if the example scripts are associated with an article too, those should have the citation as well, so that your group can get credit for the work somehow. You should also add yourself to the `AUTHORS` file in the main directory.

**Comment by speth** [comment]
Am I correct in thinking that this PR is superseded by #534 and can be closed?

**Comment by axbriones** [comment]
Ray,



This is correct.



Thanks.



Alex

________________________________

**Pull request closed without merging**
